### PR TITLE
New function `mapContainsKeyLike`

### PR DESCRIPTION
**Pull request by lingtaolf**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):  
New function` mapContainsKeyLike` to get the map that key matches a simple regular expression.
  
**Usage**:
```
SELECT * FROM map_table WHERE mapContainsKeyLike(map, '%pattern%') = 1;
```

### CODE REVIEW

File: src/Functions/map.cpp [hunk 1]
```diff
@@ -310,6 +312,123 @@ class FunctionMapValues : public IFunction
...
+        const auto & column_tuple = typeid_cast<const ColumnTuple &>(column_array.getData());
+
+        const ColumnString * column_string = checkAndGetColumn<ColumnString>(column_tuple.getColumn(0));
+        const ColumnFixedString * column_fixed_string = checkAndGetColumn<ColumnFixedString>(column_tuple.getColumn(0));
+
+        FunctionLike func_like;
+
+        for (size_t row = 0; row < input_rows_count; row++)
+        {
+            size_t element_start_row = row !=0 ? column_array.getOffsets()[row-1] : 0;
```
Comment (kssenii) [inline src/Functions/map.cpp:350]:
```suggestion
            size_t element_start_row = row != 0 ? column_array.getOffsets()[row-1] : 0;
```

File: src/Functions/map.cpp [hunk 2]
```diff
@@ -310,6 +312,123 @@ class FunctionMapValues : public IFunction
...
+            size_t elem_size = column_array.getOffsets()[row]- element_start_row;
+
+            //The keys of one row map will be processed as a single ColumnString
+            ColumnPtr sub_map_column;
+            DataTypePtr data_type;
+
+            if (column_string)
+            {
+                const ColumnString * map_key_column = checkAndGetColumn<ColumnString>(&*(column_string->cut(element_start_row, elem_size).get()));
```
Comment (kssenii) [inline src/Functions/map.cpp:359]: hm, do we really need `&*` here?

File: src/Functions/map.cpp [hunk 3]
```diff
@@ -310,6 +312,123 @@ class FunctionMapValues : public IFunction
...
+                        column,
+                        data_type,
+                        ""
+                    },
+                    arguments[1]
+                };
+
+            auto res = func_like.executeImpl(new_arguments, result_type, input_rows_count);
+
+            const auto & container = checkAndGetColumn<ColumnUInt8>(&*res)->getData();
```
Comment (kssenii) [inline src/Functions/map.cpp:385]: and here

File: src/Functions/map.cpp (hunk 1 above)
Comment (lingtaolf) [reply src/Functions/map.cpp]: done

File: src/Functions/map.cpp (hunk 2 above)
Comment (lingtaolf) [reply src/Functions/map.cpp]: updated

File: src/Functions/map.cpp (hunk 3 above)
Comment (lingtaolf) [reply src/Functions/map.cpp]: updated

**Comment by kssenii** [comment]
@lingtaolf 
Fuzzer failures are related to changes, for example trace from msan fuzzer:
```
SUMMARY: MemorySanitizer: use-of-uninitialized-value obj-x86_64-linux-gnu/../src/Common/Allocator.h:264:13 in Allocator<false, false>::checkSize(unsigned long)
Exiting
==39==WARNING: MemorySanitizer: use-of-uninitialized-value
2021.09.30 07:45:06.573693 [ 47 ] {} <Trace> BaseDaemon: Received signal -3
2021.09.30 07:45:06.574304 [ 317 ] {} <Fatal> BaseDaemon: ########################################
2021.09.30 07:45:06.575070 [ 317 ] {} <Fatal> BaseDaemon: (version 21.11.1.8254, build id: 03AF221F1F8CEF2AA089B39A933AF8B980AF0346) (from thread 313) (query_id: 9da806bd-fb8c-48e4-82b0-b7b2616b6709) Received signal Unknown signal (-3)
    #0 0x9c9a5d2 in Allocator<false, false>::checkSize(unsigned long) obj-x86_64-linux-gnu/../src/Common/Allocator.h:264:13
    #1 0x31e10966 in Allocator<false, false>::alloc(unsigned long, unsigned long) obj-x86_64-linux-gnu/../src/Common/Allocator.h:94:9
    #2 0x31e10966 in void DB::PODArrayBase<8ul, 4096ul, Allocator<false, false>, 15ul, 16ul>::alloc<>(unsigned long) obj-x86_64-linux-gnu/../src/Common/PODArray.h:120:65
    #3 0x31e10966 in DB::PODArrayBase<8ul, 4096ul, Allocator<false, false>, 15ul, 16ul>::alloc_for_num_elements(unsigned long) obj-x86_64-linux-gnu/../src/Common/PODArray.h:114:9
    #4 0x31e10966 in DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul>::PODArray(unsigned long const*, unsigned long const*) obj-x86_64-linux-gnu/../src/Common/PODArray.h:344:15
    #5 0x31e10966 in DB::ColumnString::ColumnString(DB::ColumnString const&) obj-x86_64-linux-gnu/../src/Columns/ColumnString.cpp:31:5
    #6 0x13e6f233 in DB::(anonymous namespace)::FunctionMapContainsKeyLike::executeImpl(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long) const map.cpp
    #7 0x13e79897 in DB::FunctionToExecutableFunctionAdaptor::executeImpl(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long) const (/workspace/clickhouse+0x13e79897)
```
You can also see these stack traces in details from checks.

The query which caused it is also in log `SELECT id, map FROM map_containsKeyLike_test WHERE mapContainsKeyLike(map, '1-%') = 1`, see details in check for more information.

**Comment by kssenii** [comment]
@lingtaolf hi, will you finish this and fix tests or you need help?

**Comment by lingtaolf** [comment]
> @lingtaolf hi, will you finish this and fix tests or you need help?

@kssenii Sorry! I am a little busy recently, I will check the problem tomorrow.

**Comment by lingtaolf** [comment]
@kssenii Hi ! I'm so sorry. I tried to solve the problem by myself. But I am a new comer of clickhouse development. I'm not sure what the problem is. So I think I need your help.

**Comment by kssenii** [comment]
@Mergifyio update

**Review by kssenii — APPROVED** [review]

**Pull request merged**